### ISSUE TITLE
Stale endpoints may cause istio-pilot to be stuck

### ISSUE THREAD

**Issue by phsiao** [istio/istio#25112]
**Bug description**

We are running 1.4.5-10.

We have a service that has 300+ replicas, and in the process of doing a rolling update, we would see istio-proxy to not come up with proper ingress/egress port bound as specified in the matching sidecar object.

Looking at the pilot log, right after the rolling update we will start seeing these in the log

```
2020-06-29T19:34:26.897583Z	warn	failed to get pod web/noncanary-69bc8b644d-swk5v from kube-apiserver: pods "noncanary-69bc8b644d-swk5v" not found
2020-06-29T19:34:26.897612Z	error	Endpoint without pod 10.20.214.31 noncanary.web
2020-06-29T19:34:27.114441Z	warn	failed to get pod web/noncanary-69bc8b644d-2hs9d from kube-apiserver: pods "noncanary-69bc8b644d-2hs9d" not found
2020-06-29T19:34:27.114470Z	error	Endpoint without pod 10.20.214.36 noncanary.web
2020-06-29T19:34:27.303719Z	warn	failed to get pod web/noncanary-69bc8b644d-cvnsx from kube-apiserver: pods "noncanary-69bc8b644d-cvnsx" not found
2020-06-29T19:34:27.303749Z	error	Endpoint without pod 10.20.214.4 noncanary.web
2020-06-29T19:34:27.497450Z	warn	failed to get pod web/noncanary-69bc8b644d-bsvmp from kube-apiserver: pods "noncanary-69bc8b644d-bsvmp" not found
2020-06-29T19:34:27.497483Z	error	Endpoint without pod 10.20.215.123 noncanary.web
2020-06-29T19:34:27.707925Z	warn	failed to get pod web/noncanary-69bc8b644d-f85fl from kube-apiserver: pods "noncanary-69bc8b644d-f85fl" not found
2020-06-29T19:34:27.707962Z	error	Endpoint without pod 10.20.215.222 noncanary.web
```

And these errors repeat for the same endpoint

```
2020-06-29T19:35:11.240233Z	warn	failed to get pod web/noncanary-69bc8b644d-swk5v from kube-apiserver: pods "noncanary-69bc8b644d-swk5v" not found
2020-06-29T19:35:11.240249Z	error	Endpoint without pod 10.20.214.31 web-sm.web
2020-06-29T19:35:54.905782Z	warn	failed to get pod web/noncanary-69bc8b644d-swk5v from kube-apiserver: pods "noncanary-69bc8b644d-swk5v" not found
2020-06-29T19:35:54.905810Z	error	Endpoint without pod 10.20.214.31 noncanary.web
```

The metrics `pilot_k8s_endpoints_with_no_pods` would continue to increase to 10x or more the number of endpoints this deployment has.   The metrics increments when pilot can't lookup a pod for whatever reason, see https://github.com/istio/istio/blob/a78680cc3f0c811c3f8056fd80e0ba8b9f487f5d/pilot/pkg/serviceregistry/kube/controller/endpoints.go#L219

We can recover somewhat by restarting all pilot instances, but some of the istio-proxy can still be missing ingress (egress seems to always bound after pilot restarts).

I suspect this is caused by low QPS/Burst to Kuberentes API server, and possibly an edge case in handling of stale endpoints that need one lookup before removing them, and causing the limited API call quota to be used on unneeded work.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

istio-proxy of a large deployment should work as expected.

**Steps to reproduce the bug**

Doing a rolling update of a large deployment (300+ replicas)

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

1.4.10

**How was Istio installed?**

hem/custom

**Environment where bug was observed (cloud vendor, OS, etc)**

Baremetal on-prem Kubernetes cluster

**Comment by howardjohn** [comment]
I have seen a few similar issues about this (https://github.com/istio/istio/issues/24950) recently - all of them are on 1.4, which is EOL: https://istio.io/latest/news/support/announcing-1.4-eol-final/. Its very plausible this is fixed on 1.5+ - any chance you can upgrade?

**Comment by phsiao** [comment]
> I have seen a few similar issues about this (#24950) recently - all of them are on 1.4, which is EOL: https://istio.io/latest/news/support/announcing-1.4-eol-final/. Its very plausible this is fixed on 1.5+ - any chance you can upgrade?

Yes, we are making a plan to upgrade to 1.6.

**Comment by phsiao** [comment]
FWIW, we run pilot with the following diff against 1.4.10

```diff --git a/pkg/kube/config.go b/pkg/kube/config.go
index 29622eb3a..cbb9dc990 100644
--- a/pkg/kube/config.go
+++ b/pkg/kube/config.go
@@ -70,6 +70,8 @@ func CreateClientset(kubeconfig, context string) (*kubernetes.Clientset, error)
        if err != nil {
                return nil, err
        }
+       c.QPS = 1000
+       c.Burst = 10000
        return kubernetes.NewForConfig(c)
 }
 
@@ -86,5 +88,7 @@ func createInterface(clusterConfig *clientcmdapi.Config) (kubernetes.Interface,
        if err != nil {
                return nil, err
        }
+       restConfig.QPS = 1000
+       restConfig.Burst = 10000
        return kubernetes.NewForConfig(restConfig)
 }
```

and we have verified that deployment scaling/rollout that used to cause problem no longer continue to.

**Comment by howardjohn** [comment]
We bumped up the QPS in 1.5, maybe that is why we only see it on 1.4? Although we only set it to 20

**Comment by sureshvis** [comment]
@howardjohn can we make this configurable and allow admin to override it. 
 K8s controller today allows us to configure based on cluster size (--kube-api-qps=100 --kube-api-burst=200)

**Comment by phsiao** [comment]
I suspect the API rate limit is only the trigger, and bumping it up just masks the issue.

With spending only a very limited amount of time reading the code, my hunch is that the endpoint reconciliation assumes lack of pod info is because pod cache is not yet updated, and it can retry at a later time.  However, when the pod is removed, it will never be able to pull the pod info and can become permanently stuck.

**Comment by howardjohn** [comment]
@phsiao definitely agree with it masking the behavior. I am fine with making it configurable for other reasons though.

when you ran into this did you see a memory leak? Like https://github.com/istio/istio/issues/24950 and https://github.com/kubernetes/kubernetes/issues/91686

if you do and can reproduce would be useful to grab a memory profile: https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#memory

Your explanation makes sense but I am not sure how to explain the memory leak if so. To me the memory leak implies that we are getting events but we are slow to process them? Or I guess maybe due to rate limiting the getPod call is slow which slows everything down actually.. that seems very possible

**Comment by phsiao** [comment]
Regarding memory leaks, I did not pay attention, but now I go back to metrics I do see rapid increase in memory usage from the discovery containers during the time of troubles.

<img width="1669" alt="Screen Shot 2020-06-30 at 7 42 30 PM" src="https://user-images.githubusercontent.com/31624/86187530-fe135100-bb09-11ea-80f6-db7988a2b824.png">

Does it still make sense for us to help investigating this issue in 1.4, or we should focus on upgrading to 1.6 first?  If this part of the code did not change much from 1.4 to 1.6, it might be easier for us to focus on what we have installed and port that to 1.6 later.

**Comment by howardjohn** [comment]
Based on your hypothesis, which seems very plausible, the issue is still present in 1.6 and just masked by the QPS. We can verify by dropping the QPS on 1.6 and reproducing I guess? As 1.4 is EOL we would not release a fix for this so ideally we shift focus to 1.6

**Comment by howardjohn** [comment]
@phsiao I reproduced in 1.6, I think your hypothesis is correct. great find!

**Comment by sureshvis** [comment]
This needs a very curated test scenario to reproduce.  We scaled (up and down) by 1K pods in 2 mins interval and can easily reproduce this issue. 

Pods get removed from the cache almost immediately since both delete timestamp and pod status has changed. 
https://github.com/istio/istio/blob/d89c6a624cfd0962e2ed25238ec671baebda0ca5/pilot/pkg/serviceregistry/kube/controller/pod.go#L94-L113 

updateEDS (https://github.com/istio/istio/blob/d89c6a624cfd0962e2ed25238ec671baebda0ca5/pilot/pkg/serviceregistry/kube/controller/pod.go#L94-L113)  now running through large subset,  by the time it gets to a particular pod/IP  (https://github.com/istio/istio/blob/d89c6a624cfd0962e2ed25238ec671baebda0ca5/pilot/pkg/serviceregistry/kube/controller/pod.go#L94-L113), the pod is deleted from both cache and from k8s API. 

This triggers the pilot to go in the loop (increasing pilot_k8s_endpoints_with_no_pods count) and eventually pilot running OOM 

eg: 2020-06-30T23:44:59.999312Z	error	Endpoint without pod 10.xxx.xx.xx node-exporter.kube-monitoring

**Comment by ramaraochavali** [comment]
> This triggers the pilot to go in the loop (increasing pilot_k8s_endpoints_with_no_pods count) and eventually pilot running OOM

Do not understand why would pilot will run in to loop - I agree that it exhausts the API Quota but do not understand why this would get in to loop.

**Comment by howardjohn** [comment]
> Do not understand why would pilot will run in to loop

I don't think its an infinite loop, its just stuck looping over a single endpoint's pod IPs and doing look ups. I think in there case they have 1k pod scale, so a loop over alll the endpoints at 5qps is going to be very very slow. And the next event around - probably the same

**Comment by phsiao** [comment]
> I don't think its an infinite loop

Agreed.  At first I also thought it was an infinite loop, but then I realized that is not the case because it can unstuck itself once I bumped out the qps.  But the control/data plane convergence time becomes difficult to predict when that happens and is very undesirable.   We have deployments that can finish in less than 3 minutes normally but ended up taking 10+ minutes waiting for those punitive lookups to complete and control plane status gets back to known so istio-proxy can be up.

**Comment by sureshvis** [comment]
<img width="1105" alt="Screen Shot 2020-07-01 at 9 55 06 AM" src="https://user-images.githubusercontent.com/12556103/86271061-1bcfcd00-bb81-11ea-9238-28157dbe87da.png">

In our case, it didn't get unstuck and we say "Endpoint without pod" error continued until we finally kill the pilot pod. Above the graph for "Endpoint without pod" error from pilot log.

**Comment by gotwarlost** [comment]
(I was looking at the 1.4.x code so this might not be relevant).

What I did see in that code is that a pod lookup is done using a cache and if not found, the code makes an API server call but still refuses to see a pod as non-existent even if the API explicitly returns a "not found". IMO, if the API returns an explicit "not found" pilot should remove all traces of the pod and associated endpoints from its internal structures.

**Comment by phsiao** [comment]
> IMO, if the API returns an explicit "not found" pilot should remove all traces of the pod and associated endpoints from its internal structures.

If the API call is not rate limited then that might be a good solution.  Unfortunately, the API QPS rate limiting created the undesirable side effect that caused this issue.  The fixes in the work both avoid making an API call to avoid falling into the same situation.

**Comment by gotwarlost** [comment]
> > IMO, if the API returns an explicit "not found" pilot should remove all traces of the pod and associated endpoints from its internal structures.
> 
> If the API call is not rate limited then that might be a good solution. Unfortunately, the API QPS rate limiting created the undesirable side effect that caused this issue. The fixes in the work both avoid making an API call to avoid falling into the same situation.

What I meant was, notwithstanding rate limiting (which would return a `429` or something), a `404` should always be honored as explicit evidence of non-existence?

**Comment by phsiao** [comment]
> What I meant was, notwithstanding rate limiting (which would return a `429` or something), a `404` should always be honored as explicit evidence of non-existence?

It is, but the issue is caused by the stalled reconciliation process due to rate limiting --- during a large deployment update it may take a while for all those inline API calls to complete and as a result, the system is stale and misbehaving.  By making those API calls the system risks to be put into this situation.  We have tried to tune the QPS to be as high as possible but it only relive it and not eliminate it.

**Comment by glix942** [comment]
We have experienced same issue as reported here with Pilot 1.4.5 and EKS 1.14, it caused by sudden traffic spike that up-scaled Istio Ingress Gateway rapidly (300+ pods) and down-scaled rapidly also after traffic spike passed. This issue caused Istio Ingress Gateway route request to destroyed pods, as well as incorrect services.

We have troubleshot the Pilot source code along with Kubernetes *client-go* package. We would like to share our findings here, thought on fixing this issue and **look for a solution that does not require us to maintain a customized Pilot version**.

**How does Pilot receive and handle event?**
During troubleshoot pilot code on our test cluster, here is how Pilot and its event handling works.
- Pilot creates separate `SharedIndexInformer` ([[source code](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L74-L92)]) to watch `Services`, `Nodes`, `Endpoints` and `Pod` events; ([source code](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L159-L189))
- Each informer has their own cache defined in `Indexer`; [[source code](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L83)]
- When an informer watched any change from Kubernetes Master, it pushes them into a `DeltaFIFO` ([source code](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/delta_fifo.go)) queue and processed by `HandleDeltas()` ([source code](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L344-L373));
- `sharedIndexInformer` will [*add*](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L344-L373), [*update*](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L355) or [*delete*](https://github.com/kubernetes/client-go/blob/kubernetes-1.14.9/tools/cache/shared_informer.go#L366) its own cache before distribute event to Pilot to be handled. For `endpoints` event, Pilot will then call [updateEDS()](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L904-L988) and push update to proxies; 
- Pilot event handler pushed tasks that is combination of event handler and event data into a [queue](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/queue.go), for example:
  - **endpoints** event [AddFunc](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L241)
  - **endpoints** event [UpdateFunc](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L250)
  - **endpoints** event [DeleteFunc](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L250)
- The task in the queue are executed sequentially. ([source code](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/queue.go))

**How does [updateEDS()](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L904) work out endpoints?**
Pilot will try to loop through all endpoint IP addresses presented in `endpoints` event data and try to work out correct endpoints before updating proxies
 - Pilot will invoke [c.pods.getPodByIP(ea.IP)](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/controller.go#L904) to verify corresponding Pod is presented in [informer's cache](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/pod.go#L163). **For any events, include Pos deleted event, the informer's cached is updated before the event data even distributed Pilot to handle.**
```go
		for _, ss := range ep.Subsets {
			for _, ea := range ss.Addresses {
				pod := c.pods.getPodByIP(ea.IP)
				if pod == nil {
```
- If the pod does not exist in [informer's cache](https://github.com/istio/istio/blob/1.4.5/pilot/pkg/serviceregistry/kube/controller/pod.go#L163) Pilot will invoke Kubernetes API to verify Pod existence to avoid eventual consistency.

**When does this issue happen?**
- When a Pod is created, the *pod add* event is received by information and it comes earlier than its IP address appeared in an *endpoint update* event, this makes the up-scale large amount of Pods without issue;
- When a Pod is destroyed, the *pod delete* event and *endpoint update* event are are received for each pod destroyed, and *pod delete* event comes later than corresponding *endpoint update* event. This contributed **this issue only occurs when there are large amount of Pods (150+ for our cluster) are destroyed rapidly**.

**Why this issue caused Ingress Gateway route request to destroyed pod or incorrect service?**
- When large amount of Istio Ingress Gateway pod destroyed by HPA, the *endpoints update* events created for each pod are received by Pilot and created corresponding task in its task queue and processed sequentially
- The event processing seems fine till the pod delete event arrives and cause the *endpoint update* event handler call Kubernetes Master to verify Pod existence;
- When the Kubernete API call throttled, each API call will take around 200ms, this will significantly delay the each event handling. For example in one of our tests
  - We down-scaled Ingress Gateway from 250 pods to 20 pods, there are 230 *endpoint updated* events received;
  - Pilot start invoking Kubernetes Master from the *endpoint update* event that contains 172 endpoint IP addresses;
  - The time require to process this particular event will be 172 * 200ms ~=35s; while there are still 171 Ingress gateway *endpoint updated* event in it queue. 
  - The total number of call to Kubernetes Master for Ingress Gateway pods will be  approximately 14,706 times *((n^2 - n)/2)*, that would take approximate 50 minutes;
  - At the same time, other services in cluster are up-scaling and down-scaling frequently, all the *endpoints update* events for the services will not be pushed to ingress gateway, and all other events will require calling Kubernetes Master for every IP address presented in *endpoint update* event.

**What is the impact?**
- Istio Ingress Gateway route request to destroyed pod;
- Istio Ingress Gateway route request to wrong service, due the the IP addresses are recycled and used by new pod created

** Another thought to fix **
- **Can endpoint update event for difference service handling concurrently**
- **Dedup endpoint updated event?**

**Comment by howardjohn** [comment]
We have 2 different approaches both fixing this - https://github.com/istio/istio/pull/25149 and https://github.com/istio/istio/pull/25161. Hopefully will be merged soon

**Comment by howardjohn** [comment]
Master is merged, 1.6 is open: https://github.com/istio/istio/pull/25579

**Comment by howardjohn** [comment]
Closing as this has been merged into master and release branch, will release in 1.6.6
